### ISSUE TITLE
ADD vallex

### ISSUE THREAD

**Issue by kyakuno** [axinc-ai/ailia-models#1236]
https://github.com/Plachtaa/VALL-E-X

**Comment by kyakuno** [comment]
当初はencdecが使われていたが別のモデルに置き換わった。

**Comment by kyakuno** [comment]
https://www.techno-edge.net/article/2023/08/28/1812.html

**Comment by kyakuno** [comment]
https://zenn.dev/kun432/scraps/ca42215a5f9bc6

**Comment by kyakuno** [comment]
純粋なボイスチェンジャーではなく、TTSにボイス特製を与えられるような構成。ボイスチェンジャー的に使う場合は、WhisperでTextに変換後に、TTSする。

**Comment by kyakuno** [comment]
pyopenjtalkのwindows向けのprebuilt binary
https://zenn.dev/syoyo/articles/4f6962247adfb6

**Comment by kyakuno** [comment]
audioのtokenizeにはencodecを使用している。
https://github.com/Plachtaa/VALL-E-X/blob/master/data/tokenizer.py
textのtokenizeにはopenjtalkとbpeを使用している。
textとaudioのtokenからdecoderループでmel spectrumを生成し、vocosで波形にしている。

**Comment by kyakuno** [comment]
原理的には学習不要で推論だけで、ボイスチェンジャー含めて実装できそう。

**Comment by kyakuno** [comment]
decoderはkv_cacheを使用しているため、whisperと同様に最大長で固定してONNXにエクスポート可能にする。

**Comment by kyakuno** [comment]
ar_decoder.onnxは605MB。

**Comment by kyakuno** [comment]
vocosはbackboneの後段のheadでstftを使っているため、後段のheadは除外してexportする必要がある。

**Comment by kyakuno** [comment]
ar_decoderとvocosはonnxに変換できた。
残りはAudioTokenizerで、
　self.codec.encode(wav.to(self.device))
が呼ばれる。

**Comment by kyakuno** [comment]
kv_cacheの最大長は暫定で1024。
https://storage.googleapis.com/ailia-models/vall-e-x/ar_decoder.onnx
https://storage.googleapis.com/ailia-models/vall-e-x/audio_embedding.onnx
https://storage.googleapis.com/ailia-models/vall-e-x/vocos.onnx

**Comment by kyakuno** [comment]
BPETokenizerはロジックだけなので、後は、サンプルコードを作ればつながりそうな気はする。

**Comment by kyakuno** [comment]
narのエクスポートで下記のエラーになる。
'aten::unflatten' to ONNX opset version 15 is not supported.

**Comment by kyakuno** [comment]
torch_nightly=torch 2.1.0.dev20230722でエクスポートできるとのこと。

**Comment by kyakuno** [comment]
pip3 uninstall torch
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu
で入れる。

**Comment by kyakuno** [comment]
nar_decoderもエクスポートできた。

https://storage.googleapis.com/ailia-models/vall-e-x/nar_decoder.onnx
https://storage.googleapis.com/ailia-models/vall-e-x/nar_predict_layer_0.onnx

**Comment by kyakuno** [comment]
dropoutが0.1なので、why_not_fast_path = f"dropout was {self.dropout}, required zero"で、torch._native_multi_head_attentionではなく、F.multi_head_attention_forwardが使われている。

dropoutを0にすると、torch._native_multi_head_attentionになり、Exporting the operator 'aten::_native_multi_head_attention' to ONNX opset version 15 is not supported.になる。

**Comment by kyakuno** [comment]
エクスポート時にverboseをすると、原因となった元コードの場所がわかると教えてもらった。

```
  %/nar_decoder/layers.0/self_attn/Unsqueeze_6_output_0 : Long(1, strides=[1], device=cpu) = onnx::Unsqueeze[onnx_name="/nar_decoder/layers.0/self_attn/Unsqueeze_6"](%/nar_decoder/layers.0/self_attn/Cast_1_output_0, %onnx::Unsqueeze_805), scope: models.vallex.VALLE::/modules.transformer.TransformerEncoder::nar_decoder/modules.transformer.TransformerEncoderLayer::layers.0/modules.activation.MultiheadAttention::self_attn
  %/nar_decoder/layers.0/self_attn/Concat_3_output_0 : Long(3, strides=[1], device=cpu) = onnx::Concat[axis=0, onnx_name="/nar_decoder/layers.0/self_attn/Concat_3"](%/nar_decoder/layers.0/self_attn/Constant_19_output_0, %/nar_decoder/layers.0/self_attn/Unsqueeze_5_output_0, %/nar_decoder/layers.0/self_attn/Unsqueeze_6_output_0), scope: models.vallex.VALLE::/modules.transformer.TransformerEncoder::nar_decoder/modules.transformer.TransformerEncoderLayer::layers.0/modules.activation.MultiheadAttention::self_attn # /usr/local/lib/python3.11/site-packages/torch/nn/functional.py:5346:0
  %/nar_decoder/layers.0/self_attn/Reshape_4_output_0 : Float(*, *, *, strides=[1024, 64, 1], requires_grad=0, device=cpu) = onnx::Reshape[allowzero=0, onnx_name="/nar_decoder/layers.0/self_attn/Reshape_4"](%/nar_decoder/layers.0/self_attn/Gather_4_output_0, %/nar_decoder/layers.0/self_attn/Concat_3_output_0), scope: models.vallex.VALLE::/modules.transformer.TransformerEncoder::nar_decoder/modules.transformer.TransformerEncoderLayer::layers.0/modules.activation.MultiheadAttention::self_attn # /usr/local/lib/python3.11/site-packages/torch/nn/functional.py:5346:0
  %/nar_decoder/layers.0/self_attn/Transpose_3_output_0 : Float(*, *, *, strides=[64, 1024, 1], requires_grad=0, device=cpu) = onnx::Transpose[perm=[1, 0, 2], onnx_name="/nar_decoder/layers.0/self_attn/Transpose_3"](%/nar_decoder/layers.0/self_attn/Reshape_4_output_0), scope: models.vallex.VALLE::/modules.transformer.TransformerEncoder::nar_decoder/modules.transformer.TransformerEncoderLayer::layers.0/modules.activation.MultiheadAttention::self_attn # /usr/local/lib/python3.11/site-packages/torch/nn/functional.py:5346:0
  %/nar_decoder/layers.0/self_attn/Constant_20_output_0 : Long(1, strides=[1], requires_grad=0, device=cpu) = onnx::Constant[value={351}, onnx_name="/nar_decoder/layers.0/self_attn/Constant_20"](), scope: models.vallex.VALLE::/modules.transformer.TransformerEncoder::nar_decoder/modules.transformer.TransformerEncoderLayer::layers.0/modules.activation.MultiheadAttention::self_attn
```

**Comment by kyakuno** [comment]
原因のコード。

```
    if static_k is None:
        k = k.view(k.shape[0], bsz * num_heads, head_dim).transpose(0, 1)
```

**Comment by kyakuno** [comment]
下記を書き換えるとReshapeの問題を修正できた。

```
    if static_k is None:
        #k = k.view(k.shape[0], bsz * num_heads, head_dim).transpose(0, 1) # original
        k = k.view(src_len, bsz * num_heads, head_dim).transpose(0, 1)
```

```
    if static_v is None:
        #v = v.view(v.shape[0], bsz * num_heads, head_dim).transpose(0, 1)
        v = v.view(src_len, bsz * num_heads, head_dim).transpose(0, 1)
```

**Comment by kyakuno** [comment]
RTX3080でkv_cacheの固定・非固定の速度差が大きいのが少し気になる。（処理時間は学習・推論）

kv_cache固定
processing time 829ms + 19157ms
kv_cache非固定
processing time 546ms + 1985ms

**Comment by kyakuno** [comment]
TokenEmbeddingのweightもONNXにエキスポートが必要。

**Comment by kyakuno** [comment]
PositionEmbeddingにも学習可能パラメータのalphaが含まれるのでダンプする必要がある。

**Comment by kyakuno** [comment]
prefix_modeは常に1が使われる。学習時の設定の予感。

**Comment by kyakuno** [comment]
macOSだとVALL-E EOS [240 -> 479]のトークン数で、torch = 40353ms、ailia = 33921ms。

**Comment by kyakuno** [comment]
ailia SDK 1.2.16 betaだとVALL-E EOS [240 -> 446]のトークン数でailia = 32071ms。
1.2.15と1.2.16は同じくらいの速度。

**Comment by kyakuno** [comment]
RTX3080だとTorchが1985msに対して、ailiaが28384msで速度差が大きい。
nar_decoderは高速みたいなので、ar_decoderに何か問題がありそう。

**Comment by kyakuno** [comment]
```
====Profile(Grouped by LayerType)====
LayerType       TotalPredictTime(Average)[us]   TimeRatio[%]
MatMul_DNN      36957   29.76
Gather  14913   12.01
ScatterND       13808   11.12
Transpose_DNN   12095   9.74
Slice   10533   8.48
Eltwise_DNN     10262   8.26
Slice_DNN       3777    3.04
Concat  3183    2.56
MeanVarianceNormalization_DNN   2938    2.37
Eltwise 2511    2.02
Unsqueeze       2272    1.83
Expand_DNN      1907    1.54
Expand  1774    1.43
Softmax_DNN     1710    1.38
Reshape_DNN     1000    0.81
Where   995     0.80
ReLU_DNN        987     0.79
Reshape 868     0.70
Shape   661     0.53
Range   515     0.41
Squeeze 453     0.36
Gather_DNN      74      0.06
====Profile(Summary)====
Predict Average Time[us]:124195 Variance:938637217      N:201
```

**Comment by kyakuno** [comment]
RTX3080のtorch単体で、kv_cacheの固定化をしないと1985msで、kv_cacheの固定をすると2233ms。

**Comment by kyakuno** [comment]
毎回、CUDAのハンドルの作り直しが発生している？

**Comment by kyakuno** [comment]
まだtorch依存を外せていないために競合しているのが要因な可能性もある。

**Comment by kyakuno** [comment]
onnxruntime gpuだと12510msで走る。

**Comment by kyakuno** [comment]
maskのshapeが毎フレーム変化するのが要因かもしれない。

**Comment by kyakuno** [comment]
multinominalとargmaxの違いについて。multinominalは確率分布に応じてサンプリングする。
https://stackoverflow.com/questions/61581774/what-is-the-point-of-multinomial-vs-argmax-evaluation-of-accuracy

**Comment by kyakuno** [comment]
numpyのmultinominal。
https://runebook.dev/ja/docs/numpy/reference/random/generated/numpy.random.multinomial
